terraform/eval_for_each.go

```diff
@@ -0,0 +1,64 @@
+package terraform
+
+import (
+	"fmt"
+
+	"github.com/hashicorp/hcl2/hcl"
+	"github.com/hashicorp/terraform/tfdiags"
+	"github.com/zclconf/go-cty/cty"
+)
+
+// evaluateResourceForEachExpression interprets a "for_each" argument on a resource.
+//
+// Returns a cty.Value map, and diagnostics if necessary
+func evaluateResourceForEachExpression(expr hcl.Expression, ctx EvalContext) (forEach map[string]cty.Value, diags tfdiags.Diagnostics) {
+	if expr == nil {
+		return nil, nil
+	}
+
+	forEachVal, forEachDiags := ctx.EvaluateExpr(expr, cty.DynamicPseudoType, nil)
+	diags = diags.Append(forEachDiags)
+	if diags.HasErrors() {
+		return nil, diags
+	}
+
+	if forEachVal.IsNull() {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid for_each argument",
+			Detail:   `The given "for_each" argument value is null. A map, or set of strings is allowed.`,
+			Subject:  expr.Range().Ptr(),
+		})
+		return nil, diags
+	}
+
+	if !forEachVal.CanIterateElements() || forEachVal.Type().IsListType() {
+		diags = diags.Append(&hcl.Diagnostic{
+			Severity: hcl.DiagError,
+			Summary:  "Invalid for_each argument",
+			Detail:   fmt.Sprintf(`The "for_each" argument must be a map, or set of strings.`),
+			Subject:  expr.Range().Ptr(),
+		})
+		return nil, diags
+	}
+
+	if forEachVal.Type().IsSetType() {
+		// QUESTION: Can I get one elem out of the set to check this, or do I need
+		// to do the whole iterator business. This only looks at the first item before it returns
```

Comment:
`Set` is a collection type kind, so all elements are required to have the same type, and that element type is part of the set type.

So you can write this as something like:

```go
if forEachVal.Type().IsSetType() {
    if forEachVal.Type().ElementType() != cty.String {
        // (error message)
    }
    return forEachVal.AsValueMap(), nil
}
```

Reply:
Perfect, thank you!